osm2streets-py/src/lib.rs

```diff
@@ -0,0 +1,454 @@
+                .map_err(|e| PyErr::new::<pyo3::exceptions::PyValueError, _>(format!("{}", e)))?;
+
+        let mut transformations = Transformation::standard_for_clipped_areas();
+        if input.dual_carriageway_experiment {
+            transformations.retain(|t| !matches!(t, Transformation::CollapseShortRoads));
+            transformations.push(Transformation::MergeDualCarriageways);
+        }
+        if input.sidepath_zipping_experiment {
+            transformations.push(Transformation::ZipSidepaths);
+            transformations.push(Transformation::CollapseDegenerateIntersections);
+        }
+
+        if input.debug_each_step {
+            street_network.apply_transformations_stepwise_debugging(transformations, &mut timer);
+        } else {
+            street_network.apply_transformations(transformations, &mut timer);
+        }
+
+        Ok(Self {
+            inner: street_network,
+            ways: doc.ways,
+        })
+    }
+
+    /// Converts the entire `StreetNetwork` to a GeoJSON format.
+    ///
+    /// Returns a GeoJSON string representing all elements in the street network.
+    pub fn to_geojson_plain(&self) -> PyResult<String> {
+        self.inner
+            .to_geojson(&Filter::All)
+            .map_err(|e| PyErr::new::<pyo3::exceptions::PyRuntimeError, _>(format!("{}", e)))
+    }
+
+    /// Converts lane polygons in the `StreetNetwork` to a GeoJSON format.
+    ///
+    /// Returns a GeoJSON string representing the polygons of each lane.
+    pub fn to_lane_polygons_geojson(&self) -> PyResult<String> {
+        self.inner
+            .to_lane_polygons_geojson(&Filter::All)
+            .map_err(|e| PyErr::new::<pyo3::exceptions::PyRuntimeError, _>(format!("{}", e)))
+    }
+
+    /// Converts lane markings in the `StreetNetwork` to a GeoJSON format.
+    ///
+    /// Returns a GeoJSON string representing the lane markings, such as dashed or solid lines.
+    pub fn to_lane_markings_geojson(&self) -> PyResult<String> {
+        self.inner
+            .to_lane_markings_geojson(&Filter::All)
+            .map_err(|e| PyErr::new::<pyo3::exceptions::PyRuntimeError, _>(format!("{}", e)))
+    }
+
+    /// Converts intersection markings in the `StreetNetwork` to a GeoJSON format.
+    ///
+    /// Returns a GeoJSON string representing the markings at intersections.
+    pub fn to_intersection_markings_geojson(&self) -> PyResult<String> {
+        self.inner
+            .to_intersection_markings_geojson(&Filter::All)
+            .map_err(|e| PyErr::new::<pyo3::exceptions::PyRuntimeError, _>(format!("{}", e)))
+    }
+
+    /// Retrieves debugging steps for each modification applied to the `StreetNetwork`.
+    ///
+    /// Returns a vector of `PyDebugStreets` objects, which represent intermediate states or steps
+    /// in the network transformation process.
+    pub fn get_debug_steps(&self) -> PyResult<Vec<PyObject>> {
+        Python::with_gil(|py| {
```

Comment:
Why is the GIL necessary here? A comment could be useful

Reply:
Hey Dustin, 
thank you for your feedback. Made me smile to see that everything worked as expected on your end as well :D Sorry for the late reply, I was at a conference and on the move for the last 3 weeks - now finally getting to it and resolving some of the comments. As for the need for GIL:

The GIL is needed here because Python’s CPython implementation uses it to ensure thread safety when working with Python objects. In the ```get_debug_steps``` function, we’re creating Python objects with ```into_py(py)```, and this requires holding the GIL to avoid unsafe behavior when Rust interacts with Python.

Addressed in commit d848b32. See the updated comment at [line 153-156](https://github.com/lukasbeuster/osm2streets/blob/python_wrap/osm2streets-py/src/lib.rs#L153-L156)